### ISSUE TITLE
SegmentsView Control

### ISSUE THREAD

**Issue by hnabbasi** [xamarin/Xamarin.Forms#6904]
## Summary
Have a segmented control for Xamarin.Forms. iOS has `UISegmentedControl`, but Android does not. This control will provide native looking segments in tab form for Android platform.

iOS | Android
--|--
<img src="https://raw.githubusercontent.com/hnabbasi/Segments/master/media/ios_segments_1.gif"/>  | <img src="https://raw.githubusercontent.com/hnabbasi/Segments/master/media/android_segments_1.gif" />
<img src="https://raw.githubusercontent.com/hnabbasi/Segments/master/media/ios_segments_2.gif"/>  | <img src="https://raw.githubusercontent.com/hnabbasi/Segments/master/media/android_segments_2.gif" />

## API
The `SegmentsView` implements `IViewContainer<Segment>` where `Segment` is a simple view that shows up on the UI.

`SegmentsView`:
```csharp
class SegmentsView : View, IViewContainer<Segment>, IBorderElement {}
```

> Note:
>
> `IBorderElement` can be used on iOS for `CornerRadius` and `BorderWidth` using
> `IPlatformElementConfiguration`. Android does not have border (yet).

`Segment`:
```csharp
class Segment : View {}
```

## Properties
**SegmentsView**

Name | Type | Description
---|---|---
TintColor | Xamarin.Forms.Color | Main color for the control (background, border, etc.)
Children | IList\<Segment\> | `IViewContainer<Segment>` implementation to hold `Segment`
SelectedIndexChanged | Event | Event handler for when a segment is selected. Raises `SelectedItemChangedEventArgs` with segment and selected index
UnselectedTintColor | Xamarin.Forms.Color | Background color for the unselected segment
SelectedTextColor | Xamarin.Forms.Color | Text color of the selected segment
UnselectedTextColor | Xamarin.Forms.Color | Text color of the unselected segment
SelectedIndex | System.Int32 | Index of selected segment

**Segment**

Name | Type | Description
---|---|---
Title | System.String | Title of the segment                                                            

## Usage

**XAML**
```xml
<SegmentsView SelectedIndexChanged="OnSegmentSelected">
    <SegmentsView.Children>
        <Segment Title="View A"/>
        <Segment Title="View B"/>
    </SegmentsView.Children>
</SegmentsView>
```
**C#**
```csharp
var segments = new SegmentView
    {
        Children = new List<Segment>
        {
            { new Segment { Title = "View A" } },
            { new Segment { Title = "View B" } }
        }
    };
segments.SelectedIndexChanged += OnSegmentSelected;

// handler
void OnSegmentSelected(object sender, SelectedItemChangedEventArgs e){ }
```

### iOS
On iOS, this is simply a `UISegmentedControl`.

### Android
On Android, this is be a `RadioGroup` with a custom `RadioButton` to make it look like an `ActionTab` that Android users are used to and expect.

```csharp
class FormsSegmentsView : Android.Widget.RadioGroup
```

### UWP
👀 

## Initial Implementation
I initially implemented this control inspired by Alex Rainman's [SegmentedControl](https://github.com/alexrainman/SegmentedControl), but decided to enhance the Android version to match the native experience.

Sample code: [https://github.com/hnabbasi/Segments](https://github.com/hnabbasi/Segments)

iOS | Android
--|--
<img src="https://raw.githubusercontent.com/hnabbasi/Segments/master/media/iOS.png"/> | <img src="https://raw.githubusercontent.com/hnabbasi/Segments/master/media/Android.png" />

## Enhancements
Some of the enhancements that are not part of current implementation include,
- Support binding for `Children`
- Icons in `Segment` item
- Material implementation !?

## Feedback
I have started to port this control from the sample code above to Xamarin.Forms following `CheckBox` as example. You can follow progress over at [hnabbasi/Xamarin.Forms](https://github.com/hnabbasi/Xamarin.Forms). 

I would love any feedback from you (the XF team and community) on,
- do you even need this
- does it look like something you/your users would expect on their platform
- and making it better before it is anywhere near PR-ready 😄 

Enjoy! 👍

**Comment by Tillman32** [comment]
1) As someone who has written their own control to do this already, YES I think it is useful. 
2) iOS looks totally native/expected - I really like the Android implementation too, I just cant speak to if its expected or not. 

Nice work! +1

**Comment by pranshu-aggarwal** [comment]
I want this but it would be nice if android one also looks the same as iOS

**Comment by RLittlesII** [comment]
Correct me if I am wrong, if you wanted to create and effect or renderer to make it look consistent I'd wager you can.  I actually like that ootb Android looks more Droid native.

**Comment by hnabbasi** [comment]
> I want this but it would be nice if android one also looks the same as iOS

That is exactly what I am trying not to do 😄 I want the Android users to get a native experience and producing iOS experience on Android is not ideal here.

However, if you want that kind of consistency on both platform for this control, check out the control built by Alex Rainman's [SegmentedControl](https://github.com/alexrainman/SegmentedControl)

**Comment by hnabbasi** [comment]
> Correct me if I am wrong, if you wanted to create and effect or renderer to make it look consistent I'd wager you can.  I actually like that ootb Android looks more Droid native.

Correct. Looking at popular apps - Twitter, Instagram, YouTube, Fitbit, etc - they all follow the same design. So the users are used to it. I added the **Android Design Inspiration** section outlining that.

**Comment by bcaceiro** [comment]
It would be great to have the same design i both platforms. I've come across the iOS style within Android design

**Comment by wveinti** [comment]
It would definitely be useful and a lot better be able to have the same design in both plataforms.

**Comment by djcparker** [comment]
To me, the examples in the android design inspiration look like standard material top tabs. 

I use both Fitbit and Twitter on Android and they don't seem like a segment control to me.

Happy to be wrong, of course 🙂

**Comment by mshwf** [comment]
I made similar control
https://youtu.be/9FZhLZuWBsE

**Comment by jamesmontemagno** [comment]
This is absolutely delightful and I like the idea of using the native UISegmentedControl on iOS and a tab look/feel on Android does seem right....

Maybe on Android we could play with something like this custom implementation:
![image](https://user-images.githubusercontent.com/1676321/61758594-3ba81f80-ad7a-11e9-99a6-13a8ed7da3ad.png)

**Comment by hnabbasi** [comment]
Yea. I am working on another sample. Will post tomorrow. It looks similar to ios with configurable border radius.

Hussain Abbasi

web: HussainAbbasi.com

blog: intelliAbb.com

tel: +1 87-ABBASI-87

________________________________

**Comment by djcparker** [comment]
Are you thinking of Material as well?  Would the MaterialButtonToggleGroup help with that?  
It seems to be the equivalent of ios UISegmentedControl?

**Comment by hnabbasi** [comment]
> Are you thinking of Material as well? Would the MaterialButtonToggleGroup help with that?
> It seems to be the equivalent of ios UISegmentedControl?

Yes. I love MaterialButtonToggleGroup and plan to try it out as well 👍

**Comment by jamesmontemagno** [comment]
Ahhh yeah perhaps we have a Material version of it that uses that! I couldn't find it, but that looks like what we want ;)

**Comment by hnabbasi** [comment]
iOS coming along pretty well. I think that's all I am going to do with that for now. 
Implemented [IBorderElement](https://github.com/xamarin/Xamarin.Forms/blob/master/Xamarin.Forms.Core/IBorderElement.cs) which gives us,
```csharp
Color BorderColor { get; }
int CornerRadius { get; }
Color BackgroundColor { get; }
double BorderWidth { get; }
```
Sample,
<img src="https://raw.githubusercontent.com/hnabbasi/Segments/master/media/ios_progress_1.gif">
You can follow along here [https://github.com/hnabbasi/Segments](https://github.com/hnabbasi/Segments)

Next up, Android 😃 🤖

**Comment by hnabbasi** [comment]
Android progress,
Since most of you seems to like the iOS feel for this control, here's what I have so far. Please provide feedback.

<img src="https://raw.githubusercontent.com/hnabbasi/Segments/master/media/android_progress_1.png"/>

**All Caps**

<img src="https://raw.githubusercontent.com/hnabbasi/Segments/master/media/android_progress_1_caps.png"/>

**Comment by djcparker** [comment]
It's looking really good.  Great work!

I guess I'm used to Material on Android so I tend towards typography that fits that design pattern, so probably the caps version.

The corner radius looks great on both platforms 👍.

**Comment by davidbuckleyni** [comment]
Will this allow you to embed page views as the same way that app shells do just by something like

<Page:AboutPage/>
Would be handy

**Comment by jtorvald** [comment]
This issue can be closed. Seems like it has moved to Xamarin Community Toolkit and it has stalled there. So probably will go into Maui Community Toolkit.

For people that want to implement something similar with standard xamarin forms, you could use a RadioButton Group and ControlTemplate. A small code dump below

```
<!-- in resource dictionary -->
<ControlTemplate x:Key="RadioTemplate">
                    <Frame HasShadow="False" BorderColor="Transparent" CornerRadius="0" HeightRequest="40" HorizontalOptions="Fill" VerticalOptions="Start" Padding="0" Margin="0">
                        <VisualStateManager.VisualStateGroups>
                            <VisualStateGroupList>
                                <VisualStateGroup x:Name="CheckedStates">
                                    <VisualState x:Name="Normal">
                                        <VisualState.Setters>
                                            <Setter Property="BackgroundColor" Value="Gray"/>
                                        </VisualState.Setters>
                                    </VisualState>
                                    <VisualState x:Name="Checked">
                                        <VisualState.Setters>
                                            <Setter Property="BackgroundColor" Value="Green"/>
                                        </VisualState.Setters>
                                    </VisualState>
                                    <VisualState x:Name="Unchecked">
                                        <VisualState.Setters>
                                            <Setter Property="BackgroundColor" Value="Blue"/>
                                        </VisualState.Setters>
                                    </VisualState>
                                </VisualStateGroup>
                            </VisualStateGroupList>
                        </VisualStateManager.VisualStateGroups>
                        <Grid Margin="0">
                            <ContentPresenter></ContentPresenter>
                        </Grid>
                    </Frame>
                </ControlTemplate>

<!-- segmentsview with radiobuttons -->
                            <Frame BorderColor="LightGray" CornerRadius="15" BackgroundColor="Transparent" >

                                    <Grid ColumnSpacing="0" Padding="0" BackgroundColor="Transparent"
                                                    RadioButtonGroup.GroupName="GroupA"
                                                    RadioButtonGroup.SelectedValue="{Binding SelectedValue}">
                                    <Grid.ColumnDefinitions>
                                        <ColumnDefinition Width="*"/>
                                        <ColumnDefinition Width="*"/>
                                        <ColumnDefinition Width="*"/>
                                    </Grid.ColumnDefinitions>
                                        <RadioButton ControlTemplate="{StaticResource RadioTemplate}" Grid.Column="0"
                                                        Value="A"
                                                        CornerRadius="15">
                                            <RadioButton.Content>
                                                <StackLayout HorizontalOptions="Center" VerticalOptions="Center">
                                                    <Label Text="Option A"/>
                                                </StackLayout>
                                            </RadioButton.Content>
                                        </RadioButton>
                        
                                        <RadioButton ControlTemplate="{StaticResource MetricRadioTemplate}" Grid.Column="1"
                                                        Value="B"
                                                        CornerRadius="15">
                                            <RadioButton.Content>
                                                <StackLayout HorizontalOptions="Center" VerticalOptions="Center">
                                                    <Label Text="Option B"/>
                                                </StackLayout>
                                            </RadioButton.Content>
                                        </RadioButton>

                                        <RadioButton ControlTemplate="{StaticResource RadioTemplate}" Grid.Column="2"
                                                        Value="C"
                                                        CornerRadius="15">
                                            <RadioButton.Content>
                                                <StackLayout HorizontalOptions="Center" VerticalOptions="Center">
                                                    <Label Text="option C"/>
                                                </StackLayout>
                                            </RadioButton.Content>
                                        </RadioButton>


                                    </Grid>
                                </Frame>
```
